service-coordtransform/src/main/java/fi/nls/paikkatietoikkuna/coordtransform/CoordinateTransformationActionHandler.java

```diff
@@ -83,20 +197,181 @@ public void handleAction(ActionParameters params) throws ActionException {
 
         try {
             // Reuse the double array
-            CoordTransService.parseResponse(serviceResponseBytes, coords, dimension);
+            CoordTransService.parseResponse(serviceResponseBytes, coords, targetDimension);
         } catch (IllegalArgumentException e) {
             throw new ActionException(e.getMessage());
         }
 
         HttpServletResponse response = params.getResponse();
-        response.setContentType(IOHelper.CONTENT_TYPE_JSON);
         try (OutputStream out = response.getOutputStream()) {
-            writeJsonResponse(out, coords, dimension);
+            if (transformToFile){
+                String fileName = addFileExt(exportSettings.getFileName());
+                writeFile (coords, targetDimension, exportSettings, targetCrs);
+                response.setContentType(FILE_TYPE);
```

Comment:
Might be too late to set content types or to add headers when response.getOutputStream() has already been called.